### DISCUSSION TITLE
Large number of particles for PBD cloth simulation crashed

### DISCUSSION THREAD

**Discussion by SHiziw** [NVIDIA-Omniverse/PhysX#95] [Q&A] [answered]
Hi, I'm using PBD cloth for simulation large and long film, therefore I have a large number of particles in my case. But if a large number of particles are used, maybe around 10,000,000, PhysX crashed suddenly. No prompt, no error code. How can I prevent this crash happening?

**Comment by adenzler-nvidia** [comment, marked as answer]
Thanks for reaching out. We already fixed this issue with large particle counts on our development trunk, and it will be in the next major release. The background story is that we used unsigned 32bit integers for our GPU memory allocation sizes, and with 10 million particles, one contact buffer ends up being more than 4GB leading to an integer overflow. The size of that buffer is determined by `numParticles * maxNeighborhood * sizeof(float2)` - so as long as you keep that below 4GB you should be able to avoid the issue.

Which means lowering `maxNeighborhood` is one strategy to work around this, but obviously that could affect the quality of the collisions. If not all cloths need to interact with each other, splitting things up into multiple particle systems could also help.

**Reply by SHiziw** [reply]
Thanks for your kindly and professional answer! It helps a lot.

**Reply by SHiziw** [reply]
Sorry for bother you again. I tried to simulate around 800,000 cloth particles on another device, with GRID RTX6000P-4Q, 4GB GPU memory.
And I got "PxgCudaDeviceMemoryAllocator fail to allocate memory 67108864 bytes!! Result = 2". I think in this case the number of particles is not large enough to cause the GPU memory overflow, but it still cashed.  How can I prevent this one?

**Reply by adenzler-nvidia** [reply]
This error just means you ran out of GPU memory. Are you only simulating these particles or do you have other actors in your scene?

**Reply by SHiziw** [reply]
Not much else, only 1 static rigid, 5 dynamic rigid and 5 D6Joint. I can get an out of GPU memory warning or prompt before the simulation starting？

**Reply by adenzler-nvidia** [reply]
Unfortunately it's not that simple to just warn before simulating, as the memory requirements depend on the number of contacts found etc. I acknowledge that we could have a better error message here - we will improve the wording. Additionally, GPU VRAM is a shared resource, so 4GB might be tight depending on how much GPU memory you need to for rendering and anything other than simulation that runs on the GPU.

I will double check the memory requirements locally. Could you send me the values you use in PxSceneDesc::gpuDynamicsConfig?

**Reply by SHiziw** [reply]
In fact, only minor changes have been made, based on SnippetPDBCloth.cpp. So the value in `PxSceneDesc::gpuDynamicsConfig` keeps the same in ctor:
```
		PxgDynamicsMemoryConfig() :
		tempBufferCapacity(16 * 1024 * 1024),
		maxRigidContactCount(1024 * 512),
		maxRigidPatchCount(1024 * 80),
		heapCapacity(64 * 1024 * 1024),
		foundLostPairsCapacity(256 * 1024),
		foundLostAggregatePairsCapacity(1024),
		totalAggregatePairsCapacity(1024),
		maxSoftBodyContacts(1 * 1024 * 1024),
		maxFemClothContacts(1 * 1024 * 1024),
		maxParticleContacts(1*1024*1024),
		collisionStackSize(64*1024*1024),
		maxHairContacts(1 * 1024 * 1024){}
```
